src/field_theory/algebraic_closure.lean

```diff
@@ -0,0 +1,283 @@
+/-
+Copyright (c) 2020 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+-/
+
+import algebra.direct_limit
+import field_theory.splitting_field
+import analysis.complex.polynomial
+
+/-!
+# Algebraic Closure
+
+In this file we define the typeclass for algebraically closed fields and algebraic closures.
+We also construct an algebraic closure for any field.
+
+## Main Definitions
+
+- `is_alg_closed k` is the typeclass saying `k` is an algebraically closed field, i.e. every
+polynomial in `k` splits.
+
+- `is_alg_closure k K` is the typeclass saying `K` is an algebraic closure of `k`.
+
+- `algebraic_closure k` is an algebraic closure of `k` (in the same universe).
+
+## TODO
+
+Show that any algebraic extension embeds into any algebraically closed extension (via Zorn's lemma).
+
+## Tags
+
+algebraic closure, algebraically closed
+
+-/
+
+universes u v w
+noncomputable theory
+open_locale classical big_operators
+open polynomial
+
+variables (k : Type u) [field k]
+
+/-- Typeclass for algebraically closed fields. -/
+class is_alg_closed : Prop :=
+(splits : ∀ p : polynomial k, p.splits $ ring_hom.id k)
+
+theorem polynomial.splits' {k K : Type*} [field k] [is_alg_closed k] [field K] {f : k →+* K}
+  (p : polynomial k) : p.splits f :=
+polynomial.splits_of_splits_id _ $ is_alg_closed.splits _
+
+namespace is_alg_closed
+
+theorem of_exists_root (H : ∀ p : polynomial k, p.monic → irreducible p → ∃ x, p.eval x = 0) :
+  is_alg_closed k :=
+⟨λ p, or.inr $ λ q hq hqp,
+have irreducible (q * C (leading_coeff q)⁻¹),
+by { rw ← coe_norm_unit hq.ne_zero, exact irreducible_of_associated associated_normalize hq },
+let ⟨x, hx⟩ := H (q * C (leading_coeff q)⁻¹) (monic_mul_leading_coeff_inv hq.ne_zero) this in
+degree_mul_leading_coeff_inv q hq.ne_zero ▸ degree_eq_one_of_irreducible_of_root this hx⟩
+
+end is_alg_closed
+
+instance complex.is_alg_closed : is_alg_closed ℂ :=
+is_alg_closed.of_exists_root _ $ λ p _ hp, complex.exists_root $ degree_pos_of_irreducible hp
+
+/-- Typeclass for an extension being an algebraic closure. -/
+@[class] def is_alg_closure (K : Type v) [field K] [algebra k K] : Prop :=
+is_alg_closed K ∧ algebra.is_algebraic k K
+
+namespace algebraic_closure
+
+open mv_polynomial
+
+/-- The subtype of monic irreducible polynomials -/
+@[reducible] def monic_irreducible : Type u :=
+{ f : polynomial k // monic f ∧ irreducible f }
+
+/-- Sends a monic irreducible polynomial `f` to `f(x_f)` where `x_f` is a formal indeterminate. -/
+def eval_X_self (p : monic_irreducible k) : mv_polynomial (monic_irreducible k) k :=
+polynomial.eval₂ mv_polynomial.C (X p) p
```

Comment:
The variable in the docstring doesn't match the variable in the definition. I used `f` everywhere because that is the convention later on in the file.
```suggestion
/-- Sends a monic irreducible polynomial `f` to `f(x_f)` where `x_f` is a formal indeterminate. -/
def eval_X_self (f : monic_irreducible k) : mv_polynomial (monic_irreducible k) k :=
polynomial.eval₂ mv_polynomial.C (X f) f
```